### DISCUSSION TITLE
The Without() method does not work when the class has 1 constructor with all parameters

### DISCUSSION THREAD

**Discussion by codequality44** [AutoFixture/AutoFixture#1368] [Q&A]
I have the following class:

```
 public class Foo
    {
      public Foo(string id, 
         string name, 
         string? homeTown = null, 
         IEnumerable<string>? someCollection = null)
      {
         Id = id;
         Name = name;  
         HomeTown = homeTown;
         SomeCollection = someCollection;
      }   

      public string Id { get; set; }
      public string Name {get; set;}
      public string? HomeTown {get; set;}
      public IEnumerable<string>? SomeCollection {get; set;}
    }
```
  
  I want to create a mock with AutoFixture without filling HomeTown and SomeCollection.

But when I try to create it like this, the properties still get filled!

   `Fixture fixture = new Fixture();`

    `var dto = fixture.Build<Foo>()
        .Without(x => x.HomeTown) //still gets filled in!
        .Without(x => x.SomeCollection) //still gets filled in!
        .Create();
`

Why is this behaviour?

**Comment by Kralizek** [comment]
Without delving in the code. AF finds that the class you want to generate has a constructor. It resolves a value for each parameter of the constructor and uses them to construct an instance. It doesn't have a knowledge of the connection between the constructor parameter and the property you marked to be excluded with `Without`.

You have few options:
- Add a parameterless constructor and make sure that AF doesn't use the greedy selection when selecting which constructor to use
- Use `FromFactory` to have stronger control over which parameters are used, there are overloads that accept a `Func` with multiple inputs.

You can probably have an even stronger control by creating your own specimen builder, but I'd avoid it as much as possible :)

**Reply by codequality44** [reply]
Thank you for your feedback, Kralizek!

I am thinking of filling the properties in the following way: Do you see any problems with this approach?

`var dto = fixture.Build<Foo>()
    .With(x => x.HomeTown, default(string?)) //it will fill it with null
    .With(x => x.SomeCollection, default(IEnumerable<string>?)) //it will fill it with null
    .Create();

**Reply by Kralizek** [reply]
I can't say it will work because I don't know if With is applied after the constructor.

I'd try `Build<Foo>().FromFactory((string id, string name) => new Foo(id, name)).Create()`
